modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/L0Service.scala

```diff
@@ -36,56 +40,87 @@ object L0Service {
+              val nextOrdinal = lastStoredOrdinal.map(_.next).getOrElse(latestOrdinal).value.value
+              val latestOrdinalCap = latestOrdinal.value.value
+                .min(singlePullLimit.map(nextOrdinal + _.value).getOrElse(latestOrdinal.value.value))
+
+              nextOrdinal to latestOrdinalCap
+            }
         }
+          .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
 
-      def pullGlobalSnapshots: F[List[Hashed[GlobalSnapshot]]] = {
-        for {
-          lastStoredOrdinal <- lastSnapshotStorage.getOrdinal
-          l0Peer <- l0ClusterStorage.getRandomPeer
+      def fetchMajorityGlobalSnapshot(ordinal: SnapshotOrdinal, peers: List[L0Peer]): F[Option[Hashed[GlobalSnapshot]]] =
+        peers
+          .traverse(fetchGlobalSnapshot(ordinal))
+          .map(_.flatten)
+          .map(_.sortBy(-_.signed.proofs.length).headOption)
 
-          pulled <- l0GlobalSnapshotClient.getLatestOrdinal
-            .run(l0Peer)
-            .map { lastOrdinal =>
-              val nextOrdinal = lastStoredOrdinal.map(_.next).getOrElse(lastOrdinal)
-              val lastOrdinalCap = lastOrdinal.value.value
-                .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+      def fetchGlobalSnapshot(ordinal: SnapshotOrdinal)(peer: L0Peer): F[Option[Hashed[GlobalSnapshot]]] =
+        l0GlobalSnapshotClient
+          .get(ordinal)(peer)
+          .flatMap(_.toHashedWithSignatureCheck)
+          .flatMap(_.liftTo[F])
+          .map(_.some)
+          .handleError(_ => none[Hashed[GlobalSnapshot]])
 
-              nextOrdinal.value.value to lastOrdinalCap
+      def pullGlobalSnapshots: F[List[Hashed[GlobalSnapshot]]] =
+        l0ClusterStorage.getRandomPeer.flatMap { peer =>
+          fetchOrdinals(peer).flatMap { ordinals =>
+            (ordinals, List.empty[Hashed[GlobalSnapshot]]).tailRecM {
+              case (ordinal :: nextOrdinals, snapshots) =>
+                fetchGlobalSnapshot(ordinal)(peer)
+                  .map(
+                    _.fold(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalSnapshot]])]) { s =>
+                      (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalSnapshot]]]
+                    }
+                  )
+
+              case (Nil, snapshots) =>
+                Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalSnapshot]])])
             }
-            .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
-            .flatMap { ordinals =>
+          }
+        }.handleErrorWith { e =>
+          logger.warn(e)(s"Failure pulling global snapshots!") >>
+            Applicative[F].pure(List.empty[Hashed[GlobalSnapshot]])
+        }
+
+      def pullMajorityGlobalSnapshots: F[List[Hashed[GlobalSnapshot]]] =
+        l0ClusterStorage.getPeers
+          .map(_.toNonEmptyList.toList)
+          .flatMap { peers =>
```

Comment:
Why not keep the `Nel` here? 🤔

Reply:
I'll change that, it was from before refactoring